File: projects/engineering.purdue.edu/kak/programmingwithobjects/pwocode/chap11/Robot.cc
// This code example is from the following source:
//
// Book Title:  Programming with Objects, A Comparative Presentation
//              of Object-Oriented Programming with C++ and Java
//
// Book Author:   Avinash Kak
//
// Chapter:     Chapter 11  Classes, The Rest Of The Story
//
// Section:     Section 11.3  Static Members In C++
//



//Robot.cc

#include <iostream>
#include <string>
using namespace std;

class Robot {
    int idNum;
    static int nextIdNum;                                         //(A)
    string owner;
public:
    int getIdNum();
    int whatIsNextIdNum();                                        //(B)
    string getOwner();
    Robot( string ownername);
    void print() { cout << idNum << " " << owner << endl; }
};

int Robot::nextIdNum = 1;                                         //(C)

int Robot::whatIsNextIdNum() { return nextIdNum++; }

int Robot::getIdNum() { return idNum; }

string Robot::getOwner() { return owner; }

Robot::Robot( string name ) { 
    idNum = whatIsNextIdNum();                                    //(D)
    owner = name;
}

int main()
{
    Robot r1( "ariel" );
    r1.print();                               // 1  ariel
    Robot r2( "mauriel" );
    r2.print();                               // 2  mauriel
    Robot r3( "mercurial" );
    r3.print();                               // 3  mercurial
    return 0;
}

File: projects/demos/engineering.purdue.edu/programmingwithobjects/pwocode/chap11/Robot.java
// This code example is from the following source:
//
// Book Title:  Programming with Objects, A Comparative Presentation
//              of Object-Oriented Programming with C++ and Java
//
// Book Author:    Avinash Kak
//
// Chapter:     Chapter 11  Classes, The Rest Of The Story  
//
// Section:     Section 11.4  Static Members In Java
//


//Robot.java

class Robot {
    public int idNum;
    public static int nextIdNum = 1;                              //(A)
    public String owner;

    public int getIdNum() { return idNum; }
    public String getOwner() { return owner; }

    public Robot() { idNum = nextIdNum++; };                      //(B)
    public Robot( String name) { this(); owner = name; }          //(C)

    public void print() { System.out.println( idNum + " " + owner ); }

    public static void  main( String[] args )
    {
        Robot r1 = new Robot( "ariel" );
        r1.print();                               // 1 ariel
        Robot r2 = new Robot( "mauriel" );
        r2.print();                               // 2 maurial
        Robot r3 = new Robot( "mercurial" );
        r3.print();                               // 3 mercurial
    }
}